apm-agent-plugins/apm-es-restclient-plugin/src/main/java/co/elastic/apm/es/restclient/ElasticsearchRestClientInstrumentation.java

```diff
@@ -0,0 +1,121 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.es.restclient;
+
+import co.elastic.apm.bci.ElasticApmInstrumentation;
+import co.elastic.apm.impl.transaction.Span;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.apache.http.Header;
+import org.elasticsearch.client.Request;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.ResponseException;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+/**
+ * Instrumentation for Elasticsearch RestClient, currently supporting only synchronized queries.
+ * All sync operations go through org.elasticsearch.client.RestClient#performRequest(org.elasticsearch.client.Request)
+ */
+public class ElasticsearchRestClientInstrumentation extends ElasticApmInstrumentation {
+    private static final String ES_REST_CLIENT_INSTRUMENTATION_GROUP = "elasticsearch-restclient";
+    private static final String SPAN_TYPE = "es-restclient";
+
+    private static final String REST_CLIENT_CLASS_NAME = "org.elasticsearch.client.RestClient";
+    private static final String REQUEST_CLASS_NAME = "org.elasticsearch.client.Request";
+    private static final String PERFORM_REQUEST_METHOD_NAME = "performRequest";
+
+    static final String ELASTICSEARCH_NODE_KEY = "Elasticsearch-node";
+    static final String QUERY_STATUS_CODE_KEY = "Query-status-code";
+    static final String ERROR_REASON_KEY = "Error-reason";
+
+    @Advice.OnMethodEnter
+    private static void onBeforeExecute(@Advice.Argument(0) Request request,
+                                        @Advice.Local("span") Span span,
+                                        @Advice.Local("esre") ResponseException esre) {
```

Comment:
Did this based on `net.bytebuddy.asm.Advice.Local`'s javadoc, which says:
> ... if a local variable is referenced from an exit advice method, it must also be declared by an enter advice method

Reply:
Maybe it is needed for correct instrumentation of the stack frame directives

Reply:
Sorry, I am missing something here.
> This is only needed if you want to add a @Local to the exit advice

We are inlining our advice methods, which means that there are no such methods in runtime- their compiled bytecode is injected into the instrumented method. The signature of our advice method is kind of meaningless as nobody invokes them.

> not when creating local variables which are only scoped to the exit advice

This is a directive for BB to create a local variable, I assume for the purpose of properly maintaining stack frame directives for the bytecode verifier. So both entry and exit advices are scoped together under the same method they are injected into because they are both using the same operand stack and local variable array. I believe this is the reason for this requirement.